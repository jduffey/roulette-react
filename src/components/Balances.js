import { ethers } from "ethers";
import { useState } from "react";
import { useEffect } from "react";

const provider = new ethers.providers.JsonRpcProvider("http://localhost:8545");

const TOKEN_CONTRACT_ADDRESS = "0x73511669fd4dE447feD18BB79bAFeAC93aB7F31f";

async function getEthBalance(provider, address) {
    const signer = provider.getSigner(address);
    const balance = await signer.getBalance();
    return ethers.utils.formatEther(balance);
}

// TODO leaving this for reference
// async function sendEther(provider, from, to, amount) {
//     const signer = provider.getSigner(from);
//     const tx = await signer.sendTransaction({
//         to: to,
//         value: ethers.utils.parseEther(amount)
//     });
//     return tx;
// }

async function getBlock(provider) {
    const block = await provider.getBlock();
    return block;
}

async function getTokenBalance(provider, address, tokenAddress) {
    const signer = provider.getSigner(address);
    const token = new ethers.Contract(tokenAddress, [
        "function balanceOf(address) view returns (uint)",
    ], signer);
    const balance = await token.balanceOf(address);
    const formattedBalance = ethers.utils.formatEther(balance);
    return formattedBalance;
}

async function depositEther(provider, from, tokenAddress, amount) {
    const signer = provider.getSigner(from);
    const token = new ethers.Contract(tokenAddress, [
        "function deposit() payable",
    ], signer);
    const tx = await token.deposit({
        value: ethers.utils.parseEther(amount)
    });
    return tx;
}

export function Balances() {
    const [ethBalances, setEthBalances] = useState([]);
    const [tokenBalances, setTokenBalances] = useState([]);

    const [block, setBlock] = useState({});

    const [rerender, setRerender] = useState(false);

    const [tokenSymbol, setTokenSymbol] = useState("");

    useEffect(() => {
        // Default values, including seed phrase: https://hardhat.org/hardhat-network/docs/reference
        const addresses = [
            "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266",
            "0x70997970C51812dc3A010C7d01b50e0d17dc79C8",
            "0x3C44CdDdB6a900fa2b585dd299e03d12FA4293BC",
            "0x90F79bf6EB2c4f870365E785982E1f101E93b906",
            "0x15d34AAf54267DB7D7c367839AAf71A00a2C6A65",
            "0x9965507D1a55bcC2695C58ba16FB37d819B0A4dc",
            "0x976EA74026E726554dB657fA54763abd0C3a0aa9",
            "0x14dC79964da2C08b23698B3D3cc7Ca32193d9955",
            "0x23618e81E3f5cdF7f54C3d65f7FBc0aBf5B21E8f",
            "0xa0Ee7A142d267C1f36714E4a8F75612F20a79720",
            "0xBcd4042DE499D14e55001CcbB24a551F3b954096",
            "0x71bE63f3384f5fb98995898A86B02Fb2426c5788",
            "0xFABB0ac9d68B0B445fB7357272Ff202C5651694a",
            "0x1CBd3b2770909D4e10f157cABC84C7264073C9Ec",
            "0xdF3e18d64BC6A983f673Ab319CCaE4f1a57C7097",
            "0xcd3B766CCDd6AE721141F452C550Ca635964ce71",
            "0x2546BcD3c84621e976D8185a91A922aE77ECEc30",
            "0xbDA5747bFD65F08deb54cb465eB87D40e51B197E",
            "0xdD2FD4581271e230360230F9337D5c0430Bf44C0",
            "0x8626f6940E2eb28930eFb4CeF49B2d1F2C9C1199",
        ];

        const addressWithBalancePromises = addresses.map(async (address) => {
            const balance = await getEthBalance(provider, address);
            return { address, balance };
        });

        Promise.all(addressWithBalancePromises).then((res) => {
            const newBalances = res.reduce((acc, cur) => {
                acc[cur.address] = cur.balance;
                return acc;
            }, {});
            setEthBalances(newBalances);
        });

        // TODO figure out how to get the token symbol just once.. no need to do this every time we render obviously
        const token = new ethers.Contract(
            TOKEN_CONTRACT_ADDRESS,
            ["function symbol() view returns (string)"],
            provider);
        token.symbol().then((symbol) => {
            setTokenSymbol(symbol);
        });

        const addressesWithTokenBalancePromises = addresses.map(async (address) => {
            const balance = await getTokenBalance(provider, address, TOKEN_CONTRACT_ADDRESS);
            return { address, balance };
        });

        Promise.all(addressesWithTokenBalancePromises).then((res) => {
            const newBalances = res.reduce((acc, cur) => {
                acc[cur.address] = cur.balance;
                return acc;
            }, {});
            setTokenBalances(newBalances);
        });

        getBlock(provider)
            .then((block) => {
                setBlock(block);
            });

    }, [rerender]);

    const combinedBalances = Object.keys(ethBalances).reduce((acc, address) => {
        acc[address] = {
            ethBalance: ethBalances[address],
            tokenBalance: tokenBalances[address]
        };
        return acc;
    }, {});

    return (
        <div
            style={{
                color: "white",
                fontFamily: "monospace",
                fontSize: "1.25rem",
            }}
        >
            <div>
                {
                    ["hash", "number", "timestamp"].map((key) => {
                        return (
                            <div key={key}>
                                {`${key}: ${block[key]}`}
                            </div>
                        );
                    })
                }
            </div>
            <div>
                <table>
                    <thead>
                        <tr>
                            <th>Address</th>
                            <th>ETH Balance</th>
                            <th>{tokenSymbol} Balance</th>
                            <th>Deposit 1 ETH</th>
                        </tr>
                    </thead>
                    <tbody>
                        {
                            Object.entries(combinedBalances).map(([addr, { ethBalance, tokenBalance }]) => {
                                return (
                                    <tr key={addr}>
                                        <td>{`${addr.slice(0, 6)}..${addr.slice(-4)}`}</td>
                                        <td>{Number(ethBalance).toFixed(18).padStart(18 + 6, String.fromCharCode(160))}</td>
                                        <td>{Number(tokenBalance).toFixed(18).padStart(18 + 10, String.fromCharCode(160))}</td>
                                        <td>
                                            <button
                                                onClick={() => {
                                                    depositEther(
                                                        provider,
                                                        addr,
                                                        TOKEN_CONTRACT_ADDRESS,
                                                        "1"
                                                    ).then(() => {
                                                        setRerender(!rerender);
                                                    });
                                                }}
                                            >
                                                Dep 1 ETH
                                            </button>
                                        </td>
                                    </tr>
                                );
                            })
                        }
                    </tbody>
                </table>

            </div>
        </div>
    )
}
