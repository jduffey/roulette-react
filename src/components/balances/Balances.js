import { useEffect, useState } from "react";

import {
    getEthBalance,
    getBlock,
    getTokenBalance,
    tokenSymbol,
    getPlayerSpins,
    getPlayerRewards,
    getPlayerNumberCompletionSetsCounter,
    getPlayerNumberCompletionSetCurrent,
    FIRST_PLAYER_ADDRESS,
    SECOND_PLAYER_ADDRESS,
    THIRD_PLAYER_ADDRESS,
    HOUSE_ADDRESS,
    TOKEN_CONTRACT_ADDRESS,
    ROULETTE_CONTRACT_ADDRESS,
} from "../../common/blockchainWrapper";

export function Balances() {
    const [ethBalances, setEthBalances] = useState([]);
    const [tokenBalances, setTokenBalances] = useState([]);
    const [playerRewards, setPlayerRewards] = useState([]);
    const [playerSpins, setPlayerSpins] = useState([]);
    const [playerNumberCompletionSetsCounter, setPlayerNumberCompletionSetsCounter] = useState([]);
    const [playerNumberCompletionSetCurrent, setPlayerNumberCompletionSetCurrent] = useState([]);

    const [block, setBlock] = useState({});

    const NICKNAMES = {
        [FIRST_PLAYER_ADDRESS]: "Player 1 👤",
        [SECOND_PLAYER_ADDRESS]: "Player 2 👤",
        [THIRD_PLAYER_ADDRESS]: "Player 3 👤",
        [HOUSE_ADDRESS]: "House 🏠",
        [TOKEN_CONTRACT_ADDRESS]: "Token 📜",
        [ROULETTE_CONTRACT_ADDRESS]: "Roulette 🎰",
    }

    const MS_REFRESH_INTERVAL = 2000;

    useEffect(() => {
        const interval = setInterval(() => {
            const addressesInDisplayOrder = [
                FIRST_PLAYER_ADDRESS,
                SECOND_PLAYER_ADDRESS,
                THIRD_PLAYER_ADDRESS,
                HOUSE_ADDRESS,
                TOKEN_CONTRACT_ADDRESS,
                ROULETTE_CONTRACT_ADDRESS,
            ];

            const addressWithBalancePromises = addressesInDisplayOrder.map(async (address) => {
                const balance = await getEthBalance(address);
                return { address, balance };
            });

            Promise.all(addressWithBalancePromises).then((res) => {
                const newBalances = res.reduce((acc, cur) => {
                    acc[cur.address] = cur.balance;
                    return acc;
                }, {});
                setEthBalances(newBalances);
            });

            const addressesWithTokenBalancePromises = addressesInDisplayOrder.map(async (address) => {
                const balance = await getTokenBalance(address);
                return { address, balance };
            });

            Promise.all(addressesWithTokenBalancePromises).then((res) => {
                const newBalances = res.reduce((acc, cur) => {
                    acc[cur.address] = cur.balance;
                    return acc;
                }, {});
                setTokenBalances(newBalances);
            });

            const addressesWithRewardsPromises = addressesInDisplayOrder.map(async (address) => {
                const balance = await getPlayerRewards(address);
                return { address, balance };
            });

            Promise.all(addressesWithRewardsPromises).then((res) => {
                const newBalances = res.reduce((acc, cur) => {
                    acc[cur.address] = cur.balance;
                    return acc;
                }, {});
                setPlayerRewards(newBalances);
            });

            const addressesWithSpinsPromises = addressesInDisplayOrder.map(async (address) => {
                const count = await getPlayerSpins(address);
                return { address, count };
            });

            Promise.all(addressesWithSpinsPromises).then((res) => {
                const counts = res.reduce((acc, cur) => {
                    acc[cur.address] = cur.count;
                    return acc;
                }, {});
                setPlayerSpins(counts);
            });

            const addressesWithNumberCompletionSetsCounterPromises = addressesInDisplayOrder.map(async (address) => {
                const count = await getPlayerNumberCompletionSetsCounter(address);
                return { address, count };
            });

            Promise.all(addressesWithNumberCompletionSetsCounterPromises).then((res) => {
                const counts = res.reduce((acc, cur) => {
                    acc[cur.address] = cur.count;
                    return acc;
                }, {});
                setPlayerNumberCompletionSetsCounter(counts);
            });

            const addressesWithCompletionsSetCurrentPromises = addressesInDisplayOrder.map(async (address) => {
                const currentSet = await getPlayerNumberCompletionSetCurrent(address);
                return { address, currentSet };
            });

            Promise.all(addressesWithCompletionsSetCurrentPromises).then((res) => {
                const currentSets = res.reduce((acc, cur) => {
                    acc[cur.address] = cur.currentSet;
                    return acc;
                }, {});
                setPlayerNumberCompletionSetCurrent(currentSets);
            });

            getBlock()
                .then((blockData) => {
                    setBlock(blockData);
                });
        }, MS_REFRESH_INTERVAL);

        return () => clearInterval(interval);
    }, []);

    const combinedBalances = Object.keys(ethBalances).reduce((acc, address) => {
        acc[address] = {
            ethBalance: ethBalances[address],
            tokenBalance: tokenBalances[address],
            rewards: playerRewards[address],
            spins: playerSpins[address],
            numberCompletionSetsCounter: playerNumberCompletionSetsCounter[address],
            numberCompletionSetCurrent: playerNumberCompletionSetCurrent[address],
        };
        return acc;
    }, {});

    const columnNamesAndWidths = {
        nickname: "16%",
        address: "16%",
        ethBalance: "16%",
        tokenBalance: "16%",
        rewards: "16%",
        spins: "10%",
        numberCompletionSetsCounter: "10%",
        numberCompletionSetCurrent: "16%",
    }

    return (
        <div
            className="Balances-component"
        >
            <div>
                {
                    ["hash", "number", "timestamp"].map((key) => {
                        return (
                            <div key={key}>
                                {`${key}: ${block[key]}`}
                            </div>
                        );
                    })
                }
            </div>
            <div>
                <table
                    className="balances-table">
                    <thead>
                        <tr className="balances-table-headers">
                            <th style={{ width: columnNamesAndWidths.nickname }}>Nickname</th>
                            <th style={{ width: columnNamesAndWidths.address }}>Address</th>
                            <th style={{ width: columnNamesAndWidths.ethBalance }}>ETH Balance</th>
                            <th style={{ width: columnNamesAndWidths.tokenBalance }}>{tokenSymbol} Balance</th>
                            <th style={{ width: columnNamesAndWidths.rewards }}>Rewards</th>
                            <th style={{ width: columnNamesAndWidths.spins }}>Spins</th>
                            <th style={{ width: columnNamesAndWidths.numberCompletionSetsCounter }}>✅ Sets</th>
                            <th style={{ width: columnNamesAndWidths.numberCompletionSetCurrent }}>Current Set</th>
                        </tr>
                    </thead>
                    <tbody>
                        {
                            Object.entries(combinedBalances).map(([addr, { ethBalance, tokenBalance, rewards, spins, numberCompletionSetsCounter, numberCompletionSetCurrent }]) => {
                                return (
                                    <tr key={addr}>
                                        <td>{NICKNAMES[addr]}</td>
                                        <td>
                                            {`${addr.slice(0, 5)}..${addr.slice(-3)}`}
                                        </td>
                                        <td>
                                            {Number(ethBalance).toLocaleString(undefined, {
                                                minimumFractionDigits: 8,
                                                maximumFractionDigits: 8
                                            })}
                                        </td>
                                        <td>
                                            {Number(tokenBalance).toLocaleString(undefined, {
                                                minimumFractionDigits: 2,
                                                maximumFractionDigits: 2
                                            })}
                                        </td>
                                        <td>
                                            {Number(rewards).toLocaleString(undefined, {
                                                minimumFractionDigits: 2,
                                                maximumFractionDigits: 2
                                            })}
                                        </td>
                                        <td>
                                            {Number(spins).toLocaleString(undefined, {
                                                minimumFractionDigits: 0,
                                                maximumFractionDigits: 0
                                            })}
                                        </td>
                                        <td>
                                            {Number(numberCompletionSetsCounter).toLocaleString(undefined, {
                                                minimumFractionDigits: 0,
                                                maximumFractionDigits: 0
                                            })}
                                        </td>
                                        <td>
                                            {typeof numberCompletionSetCurrent !== "undefined"
                                                ? numberCompletionSetCurrent.join(",")
                                                : "Empty"
                                            }
                                        </td>
                                    </tr>
                                );
                            })
                        }
                    </tbody>
                </table>
            </div>
        </div>
    )
}
