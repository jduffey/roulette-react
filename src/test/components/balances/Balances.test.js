import renderer from "react-test-renderer";

import { Balances } from "../../../components/balances/Balances";

describe("Balances", () => {
    it("renders", () => {
        const view = renderer.create(<Balances />);

        expect(view).toMatchSnapshot();
    });
});
