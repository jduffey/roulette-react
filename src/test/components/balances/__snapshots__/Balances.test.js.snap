// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Balances renders 1`] = `
<div
  className="Balances-component"
>
  <div>
    <div>
      hash: undefined
    </div>
    <div>
      number: undefined
    </div>
    <div>
      timestamp: undefined
    </div>
  </div>
  <div>
    <table
      className="balances-table"
    >
      <thead>
        <tr
          className="balances-table-headers"
        >
          <th>
            Nickname
          </th>
          <th>
            Address
          </th>
          <th
            className="Balances-eth-balance"
          >
            ETH Balance
          </th>
          <th
            className="Balances-token-balance"
          >
             Balance
          </th>
          <th
            className="Balances-token-balance"
          >
            Rewards
          </th>
          <th
            className="Balances-button-header"
          >
            Get Tokens
          </th>
          <th
            className="Balances-button-header"
          >
            Redeem Tokens
          </th>
          <th
            className="Balances-button-header"
          >
            Txfr Tokens
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
  </div>
</div>
`;
